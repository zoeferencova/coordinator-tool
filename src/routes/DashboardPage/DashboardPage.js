import React from 'react';
import NavBar from '../../components/NavBar/NavBar';
import Header from '../../components/Header/Header';
import ChartWrapper from '../../components/ChartWrapper/ChartWrapper';
import PieChartWrapper from '../../components/PieChartWrapper/PieChartWrapper'
import * as d3 from 'd3';
import config from '../../config'

import styles from './DashboardPage.module.css'
import { formatDefaultLocale } from 'd3';

export default class DashboardPage extends React.Component {
    state = {
        dataType: 'created_days',
        span: 'days',
        type: 'created',
        completionTime: {},
        pending: '',
        completion: '',

    } 

    componentDidMount() {
        d3.json(`${config.API_ENDPOINT}/data/time-completed-data`, { headers: { "Authorization": `Bearer ${window.sessionStorage.getItem(config.TOKEN_KEY)}` } })
            .then(res => res[0].length === 0 ? this.setState({ completionTime: { hour: "--", minute: "--"} }) : this.getHourMinute(res[0].map(item => this.getMinutes(item)).reduce((acc, el) => acc + el)/res[0].length))
            .then(item => item !== undefined ? this.setState({ completionTime: item }) : item)

        d3.json(`${config.API_ENDPOINT}/data/dashboard-data`, { headers: { "Authorization": `Bearer ${window.sessionStorage.getItem(config.TOKEN_KEY)}` } })
            .then(res => this.formatDashData(res))
    }

    getMinutes(item) {
        let minutes = 0;
        if (item.difference.months) {
            minutes += item.difference.months*4*7*24*60 
        }
        if (item.difference.weeks) {
            minutes += item.difference.weeks*7*24*60
        }
        if (item.difference.days) {
            minutes += item.difference.days*24*60 
        }
        if (item.difference.hours) {
            minutes += item.difference.hours*60
        }
        if (item.difference.minutes) {
            minutes += item.difference.minutes
        }
        if (item.difference.seconds) {
            minutes += 0
        }
        item = minutes

        return item
    }
    
    getHourMinute(item) {
        const minute = Number(item) % 60;
        const hour = Number(item)/60
        return { hour: parseInt(hour), minute: parseInt(minute) }
    }

    formatDashData(data) {
        this.setState({ pending: Number(data[0][0].pending) })
        if (Number(data[2][0].completed) === 0) {
            this.setState({ completion: 0 })
        } else {
            const percentage = (Number(data[2][0].completed) * 100)/Number(data[1][0].created);
            this.setState({ completion: parseInt(percentage) })
        }
    }

    changeType = (type) => {
        const dataType = `${type}_${this.state.span}`
        this.setState({ dataType, type })
    }

    changeSpan = (span) => {
        const dataType = `${this.state.type}_${span}`
        this.setState({ dataType, span })
    }
    
    render() {
        return (
            <div className="container">
                <main className="content">
                    <Header title={'Dashboard'} />
                    <section>
                        <p>Current pending requests: {this.state.pending}</p>
                        <p>Completion rate of new requests this week: {this.state.completion}%</p>
                        <p>Average completion time of requests completed this week: Hours: {this.state.completionTime.hour}, Minutes: {this.state.completionTime.minute}</p>
                    </section>
                    <section>
                        <h2>Requests by PM</h2>
                        <PieChartWrapper />
                    </section>
                    <section>
                        <h2>Completed Requests</h2>
                        <form>
                            <input type="radio" name="span" id="days" value="days" defaultChecked="checked" onChange={e => this.changeSpan(e.target.value)}></input>
                            <label htmlFor="days"> Day</label><br></br>
                            <input type="radio" name="span" id="weeks" value="weeks" onChange={e => this.changeSpan(e.target.value)}></input>
                            <label htmlFor="weeks"> Week</label><br></br>
                            <input type="radio" name="span" id="months" value="months" onChange={e => this.changeSpan(e.target.value)}></input>
                            <label htmlFor="months"> Month</label><br></br><br></br>
                        </form>
                        <form>
                            <input type="radio" name="type" id="created" value="created" defaultChecked="checked" onChange={e => this.changeType(e.target.value)}></input>
                            <label htmlFor="created"> Created</label><br></br>
                            <input type="radio" name="type" id="completed" value="completed" onChange={e => this.changeType(e.target.value)}></input>
                            <label htmlFor="completed"> Completed</label>
                        </form>
                        <ChartWrapper dataType={this.state.dataType} />
                    </section>
                </main>
                <NavBar className="nav" />
            </div>
        )
    }
}