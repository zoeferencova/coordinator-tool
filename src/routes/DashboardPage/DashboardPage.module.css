

.dashboard {
    display: flex;
    flex-direction: column;
}

.statSection, .chartSection {
    display: flex;
    flex-direction: column;
}

.dashboardContainer {
    background-color: white;
    padding: 20px 30px;
    border-radius: 7px;
    box-shadow: 2px 2px 40px -20px lightgray;
    margin: 15px 0px;
    display: flex;
    flex-direction: column;
    justify-content: center;
}

.dashboardContainer p {
    text-align: left;
    margin: 0;
}

.statName {
    text-align: left;
    color: gray;
    font-size: 14px;
}

.stat {
    font-size: 20px;
    font-weight: bold;
    color: gray;
}

.time {
    text-transform: uppercase;
    font-size: 12px;
    margin-left: 5px;
    margin-right: 10px;
}

.radioButton {
    display: none;
}

.radioButton:checked + .radioLabel {
    background-color: #edf1f5;
    font-weight: bold;
}

.radioLabel:hover {
    background-color: #dce0e6;
}

.radioLabel {
    display: inline-block;
    border: 1px solid #cad0db;
    padding: 5px;
    font-size: 10px;
    margin-bottom: 15px;
    transition: 500;
}

.first {
    border-top-left-radius: 8px;
    border-bottom-left-radius: 8px;
    border-right: none;
}

.last {
    border-top-right-radius: 8px;
    border-bottom-right-radius: 8px;
}

.middle {
    border-right: none;
}

.radioLabel:hover {
    cursor: pointer;
}

.radioContainer {
    display: flex;
    flex-direction: row;
    padding: 10px 0px;
    justify-content: space-between;
}

.chartContainer {
    padding: 15px;
}


@media only screen and (min-width: 850px) {

    .statContainer {
        width: 30%;
    }

    .chartContainer {
        width: 50%;
        padding: 20px 30px;
    }

    .statSection, .chartSection {
        flex-direction: row;
    }

    .radioLabel {
        font-size: 12px;
        padding: 7px;
    }

    .radioContainer {
        padding: 10px 40px;
    }

    .dashboardContainer {
        margin: 15px;
    }
}
