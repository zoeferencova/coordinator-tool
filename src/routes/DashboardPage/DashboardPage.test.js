import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import DashboardPage from './DashboardPage'

describe(`DashboardPage component`, () => {
  it('renders a .DashboardPage by default', () => {
    const wrapper = shallow(<DashboardPage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})