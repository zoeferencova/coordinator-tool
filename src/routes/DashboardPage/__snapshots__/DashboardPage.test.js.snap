// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage component renders a .DashboardPage by default 1`] = `
<div
  className="container"
>
  <main
    className="content"
  >
    <Header
      title="Dashboard"
    />
    <div
      className="dashboard"
    >
      <div
        className="statSection"
      >
        <section
          className="dashboardContainer statContainer"
        >
          <p
            className="statName"
          >
            Current pending requests
          </p>
          <p
            className="stat"
          />
        </section>
        <section
          className="dashboardContainer statContainer"
        >
          <p
            className="statName"
          >
            Completion rate of new requests
          </p>
          <p
            className="stat"
          >
            %
          </p>
        </section>
        <section
          className="dashboardContainer statContainer"
        >
          <p
            className="statName"
          >
            Average request completion time
          </p>
          <p
            className="stat"
          >
            <span
              className="time"
            >
              Hours
            </span>
             
            <span
              className="time"
            >
              Minutes
            </span>
          </p>
        </section>
      </div>
      <div
        className="chartSection"
      >
        <section
          className="dashboardContainer chartContainer"
        >
          <h2>
            Requests by PM
          </h2>
          <PieChartWrapper />
        </section>
        <section
          className="dashboardContainer chartContainer"
        >
          <h2>
            Created Requests
          </h2>
          <div
            className="radioContainer"
          >
            <form
              className="typeRadio"
            >
              <input
                className="radioButton"
                defaultChecked="checked"
                id="created"
                name="type"
                onChange={[Function]}
                type="radio"
                value="created"
              />
              <label
                className="radioLabel first"
                htmlFor="created"
              >
                 Created
              </label>
              <input
                className="radioButton"
                id="completed"
                name="type"
                onChange={[Function]}
                type="radio"
                value="completed"
              />
              <label
                className="radioLabel last"
                htmlFor="completed"
              >
                 Completed
              </label>
            </form>
            <form
              className="spanRadio"
            >
              <input
                className="radioButton"
                defaultChecked="checked"
                id="days"
                name="span"
                onChange={[Function]}
                type="radio"
                value="days"
              />
              <label
                className="radioLabel first"
                htmlFor="days"
              >
                 Day
              </label>
              <input
                className="radioButton"
                id="weeks"
                name="span"
                onChange={[Function]}
                type="radio"
                value="weeks"
              />
              <label
                className="radioLabel middle"
                htmlFor="weeks"
              >
                 Week
              </label>
              <input
                className="radioButton"
                id="months"
                name="span"
                onChange={[Function]}
                type="radio"
                value="months"
              />
              <label
                className="radioLabel last"
                htmlFor="months"
              >
                 Month
              </label>
            </form>
          </div>
          <ChartWrapper
            color="blue"
            dataType="created_days"
          />
        </section>
      </div>
    </div>
  </main>
  <NavBar
    className="nav"
  />
</div>
`;
